### PR TITLE
feat(manager/copier): Implement manager

### LINKED ISSUE #25556: copier support
# New package manager questionnaire

Did you read our documentation on adding a package manager?

- [X] I've read the [adding a package manager](adding-a-package-manager.md) documentation.

### What's the name of the package manager?
[Copier](https://copier.readthedocs.io/en/latest/)

### What language(s) does this package manager support?
It is a tool that generates/scaffolds any text files, according to the template. The templates are Git repos, and they use Git tags for versioning.

### How popular is this package manager?
Currently, 1.3k stars on GitHub, and 139 forks.

### Does this language have other (competing?) package managers?

- [X] Yes (give names).
* Cookiecutter
* Yeoman

### What are the big selling points for this package manager?

It is not for updating dependencies, but it is intended to keep the scaffolded code up-to-date.

For example, README file is part of scaffolding, and if a badge needs to be updated, the automatic re-run of copier (executed by Renovatebot) would prepare a PR to update the badge.

### What kind of package files, and names, does this package manager use?
Every project scaffolded using copier has the following file: `.copier-answers.yml`

### Which [`fileMatch`](../usage/configuration-options.md#filematch) pattern(s) should Renovate use?
Probably something like `.copier-answers.yml`

### Do many users need to extend the [`fileMatch`](../usage/configuration-options.md#filematch) pattern for custom file names?
Probably something like `(.*)\.copier-answers.yml$`

- [ ] Yes, provide details.
- [X] No.

### Is the [`fileMatch`](../usage/configuration-options.md#filematch) pattern going to get many "false hits" for files that have nothing to do with package management?

IMHO not.

### Can package files have "local" links to each other that need to be resolved?

No.

### Package file parsing method

The package files should be:

- [ ] Parsed together (in serial).
- [X] Parsed independently.

### Which format/syntax does the package file use?

- [ ] JSON
- [ ] TOML
- [X] YAML
- [ ] Custom (explain below)

### How should we parse the package files?

- [X] Off the shelf parser.
- [ ] Using regex.
- [ ] Custom-parsed line by line.
- [ ] Other.

### Does the package file have different "types" of dependencies?

- [ ] Yes, production and development dependencies.
- [X] No, all dependencies are treated the same.

### List all the sources/syntaxes of dependencies that can be extracted

The YAML file contains the following fields:
* `_commit`, typically Git tag, e.g. `v2.7.0`. [PEP 440](https://peps.python.org/pep-0440/)
* _src_path`: The template path, e.g. `git@github.com:pawamoy/copier-poetry.git`

### Describe which types of dependencies above are supported and which will be implemented in future

The template doesn't have to be a Git repo, it could be also a local path. But this is not interesting for Renovate, IMHO.

When executing copier, it uses the latest version according to PEP-440, but e.g. latest commit from a branch could be specified manually: https://copier.readthedocs.io/en/latest/generating/#copying-dirty-changes

### What versioning scheme does the package file(s) use?
[PEP 440](https://peps.python.org/pep-0440/)

### Does this versioning scheme support range constraints, like `^1.0.0` or `1.x`?

- [ ] Supports range constraints (for example: `^1.0.0` or `1.x`), provide details.
- [X] No.

### Is a new datasource required?

- [ ] Yes, provide details.
- [X] No.

### Will users want (or need to) set a custom host or custom registry for Renovate's lookup?

- [X] Yes, provide details.
- IMHO same as when updating e.g. Git submodule.
- [ ] No.

Where can Renovate find the custom host/registry?

- [ ] No custom host or registry is needed.
- [x] In the package file(s), provide details.
- IMHO same as e.g. Git submodule.
- [ ] In some other file inside the repository, provide details.
- [ ] User needs to configure Renovate where to find the information, provide details.

### Are there any constraints in the package files that Renovate should use in the lookup procedure?

- [x] Yes, there are constraints on the parent language (for example: supports only Python `v3.x`), provide details.
- copier requires Python 3.8 or newer
- [ ] Yes, there are constraints on the parent platform (for example: only supports Linux, Windows, etc.), provide details.
- [X] Yes, some other kind of constraint, provide details.
- Requires Git 2.27 or newer.
- [ ] No constraints.

### Will users need the ability to configure language or other constraints using Renovate config?

- [ ] Yes, provide details.
- [X] No.

### Does the package manager use a lock file or checksum file?

- [ ] Yes, uses lock file.
- [ ] Yes, uses checksum file.
- [ ] Yes, uses lock file _and_ checksum file.
- [X] No lock file or checksum.

### Is the locksum or checksum mandatory?

- [ ] Yes, locksum is mandatory.
- [ ] Yes, checksum is mandatory.
- [ ] Yes, lock file _and_ checksum are mandatory.
- [ ] No mandatory locksum or checksum.
- [X] Package manager does not use locksums or checksums.

#### Does the package manager use a cache?

- [ ] Yes, provide details.
- [X] No.

#### If the package manager uses a cache, how can Renovate control the cache?

- [X] Package manager does not use a cache.
- [ ] Controlled via command line interface, provide details.
- [ ] Controlled via environment variables, provide details.

#### Should Renovate keep a cache?

- [X] Yes, ignore/disable the cache.
- Same as for e.g. Git submodules.
- [ ] No.

### Generating a lockfile from scratch

Renovate can perform "lock file maintenance" by getting the package manager to generate a lockfile from scratch.
Can the package manager generate a lockfile from scratch?

- [ ] Yes, explain which command Renovate should use to generate the lockfile.
- [ ] No, the package manager does _not_ generate a lockfile from scratch.
- [X] No, the package manager does not use lockfiles.

### What else should we know about this package manager?

_Originally posted by @augi in https://github.com/renovatebot/renovate/discussions/25538#discussioncomment-7449537_

### PR DESCRIPTION
**Pull request by lkubb**
## Changes
* Adds support for updating [Copier](https://copier.readthedocs.io/en/latest/) templates
* ~~Adds one self-hosted (`copierTrust`) and one repository configuration (`copierOptions`, with several children)~~

## Context
Closes #25556 

* This implementation does not support `git@github.com:some/repo`-style URLs (I don't think they are used widely though)
* It does not fail when the smart update algorithm results in merge conflicts, they are not always reliable. It adds a warning when a file is reported to have conflicts though.

Note: This is my first PR to a TypeScript project in general and I have limited prior experience with Renovate. I read the contribution guidelines and tried to follow existing examples, but `copier` is not a regular manager, so I hope this is how it should be done.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
why do we need so many new config options? we don't have it for any other managers

**Comment by lkubb** [comment]
> why do we need so many new config options? we don't have it for any other managers

I'd argue that Copier is somewhat different from other managers because it can impact each file in the repository in relatively dynamic ways. Also honestly, I tend to lean towards providing more configurability to users when it does not add significant complexity, but am not married to most of the options.

I believe `copierTrust` (without this, some templates are unmanageable), `data` (for centrally updating some variables) and likely `recopy` (if merge conflicts should be avoided at all cost) are essential.

Options like `skip` and `exclude` can help address specific annoying quirks (in some generated projects, example files are re-added even if they have been removed from the repository or can cause merge conflicts). `dataFile` and `skipTasks` are probably expendable and could be removed if we want to simplify the configuration.

**Comment by rarkins** [comment]
We need to go back to the discussion for all of this. It's undesirable to bloat the configs with so many new options so we should discuss them one by one and reach agreement chest.

File: lib/modules/manager/types.ts [hunk 1]
```diff
@@ -41,6 +41,7 @@ export interface UpdateArtifactsConfig {
   registryAliases?: Record<string, string>;
   lockFiles?: string[];
   env?: UserEnv;
+  allowScripts?: boolean;
```
Comment (viceice) [inline lib/modules/manager/types.ts:44]: Wrong, not passed there.

File: lib/modules/manager/copier/artifacts.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+const DEFAULT_COMMAND_OPTIONS = ['--skip-answered', '--defaults'];
+
+function buildCommand(
+  config: UpdateArtifactsConfig,
+  packageFileName: string,
+  newVersion: string,
+): string {
+  const command = ['copier', 'update', ...DEFAULT_COMMAND_OPTIONS];
+  if (config?.allowScripts) {
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:21]:
```suggestion
  if (!GlobalConfig.get('allowScripts') || config.ignoreScripts) {
```

File: lib/modules/manager/copier/artifacts.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      packageFileName,
+      'Missing copier template version to update to',
+    );
+  }
+
+  const command = buildCommand(config, packageFileName, newVersion);
+  try {
+    await exec(command);
+  } catch (err) {
+    logger.error({ err }, `Failed to update copier template: ${err.message}`);
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:72]:
```suggestion
    logger.debug({ err }, `Failed to update copier template: ${err.message}`);
```

File: lib/modules/manager/copier/artifacts.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    return null;
+  }
+
+  if (status.conflicted.length > 0) {
+    // Sometimes, Copier erroneously reports conflicts.
+    const msg =
+      `Updating the Copier template yielded ${status.conflicted.length} merge conflicts. ` +
+      'Please check the proposed changes carefully! Conflicting files:\n  * ' +
+      status.conflicted.join('\n  * ');
+    logger.warn({ packageFileName, depName: updatedDeps[0]?.depName }, msg);
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:88]:
```suggestion
    logger.debug({ packageFileName, depName: updatedDeps[0]?.depName }, msg);
```

File: lib/modules/manager/copier/artifacts.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  if (!newVersion) {
+    return artifactError(
+      packageFileName,
+      'Missing copier template version to update to',
+    );
+  }
+
+  const command = buildCommand(config, packageFileName, newVersion);
+  try {
+    await exec(command);
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:70]: Needs dynamic install support

**Review by viceice — CHANGES_REQUESTED** [review]
Blocked by https://github.com/containerbase/base/issues/2870

File: lib/modules/manager/copier/artifacts.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  return [
+    {
+      artifactError: {
+        lockFile: packageFileName,
+        stderr: message,
+      },
+    },
+  ];
+}
+
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:48]: move to `./utils.ts`, will be replaced by our detect env tool soon

File: lib/modules/manager/copier/artifacts.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+function getPythonVersionConstraint(
+  config: UpdateArtifactsConfig,
+): string | undefined | null {
+  const { constraints = {} } = config;
+  const { python } = constraints;
+
+  if (python) {
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:55]:
```suggestion
  if (is.nonEmptyString(python)) {
```

File: lib/modules/manager/copier/artifacts.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  return undefined;
+}
+
+function getCopierVersionConstraint(config: UpdateArtifactsConfig): string {
+  const { constraints = {} } = config;
+  const { copier } = constraints;
+
+  if (is.string(copier)) {
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:67]: see above

File: lib/modules/manager/copier/artifacts.ts [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+  if (!newVersion) {
+    return artifactError(
+      packageFileName,
+      'Missing copier template version to update to',
+    );
+  }
+
+  const command = buildCommand(config, packageFileName, newVersion);
+  const pythonConstraint = getPythonVersionConstraint(config);
+  const copierConstraint = getCopierVersionConstraint(config);
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.ts:98]: inline variable

File: lib/modules/manager/copier/artifacts.spec.ts [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        modified: ['.copier-answers.yml'],
+        not_added: [],
+        deleted: [],
+        renamed: [],
+      }),
+    );
+  });
+
+  afterEach(() => {
+    GlobalConfig.reset();
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.spec.ts:57]:
```suggestion
```

set above will override all

File: lib/modules/manager/copier/artifacts.spec.ts [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+
+      expect(result).toEqual([
+        {
+          artifactError: {
+            lockFile: '.copier-answers.yml',
+            stderr: 'Missing copier template version to update to',
+          },
+        },
+      ]);
+      expect(execSnapshots).toMatchObject([]);
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.spec.ts:85]:
```suggestion
      expect(execSnapshots).toEqual([]);
```

File: lib/modules/manager/copier/artifacts.spec.ts [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+
+      expect(result).toEqual([
+        {
+          artifactError: {
+            lockFile: '.copier-answers.yml',
+            stderr: 'Unexpected number of dependencies: 0 (should be 1)',
+          },
+        },
+      ]);
+      expect(execSnapshots).toMatchObject([]);
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.spec.ts:106]:
```suggestion
      expect(execSnapshots).toEqual([]);
```

File: lib/modules/manager/copier/artifacts.spec.ts [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+      });
+
+      expect(execSnapshots).toMatchObject([
+        {
+          cmd: 'copier update --skip-answered --defaults --answers-file .copier-answers.yml --vcs-ref 1.1.0',
+        },
+      ]);
+    });
+
+    it.each([
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.spec.ts:126]: use string template style

File: lib/modules/manager/copier/artifacts.spec.ts [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+            config: {
+              ...config,
+              constraints: constraintConfig,
+            },
+          }),
+        ).not.toBeNull();
+
+        expect(execSnapshots).toMatchObject([
+          { cmd: 'install-tool python ' + (pythonConstraint ?? '3.12.4') },
+          { cmd: 'install-tool copier ' + (copierConstraint ?? '9.2.0') },
```
Comment (viceice) [inline lib/modules/manager/copier/artifacts.spec.ts:165]: use string template

**Review by viceice — APPROVED** [review]

**Pull request merged**